backend/src/test/scala/com/softwaremill/adopttapir/starter/StarterServiceITTest.scala

```diff
@@ -3,49 +3,206 @@ package com.softwaremill.adopttapir.starter
-    new StarterServiceInTests(config, new ProjectTemplateInTests(config))
+  private def testOption(shouldTest: Boolean, testFunction: TestFunction): Option[TestFunction] = {
+    if (shouldTest) Some(testFunction) else None
   }
+}
 
-  private def sbtCompileTest(file: IO[File]): Unit = {
-    IO.blocking(BFile.newTemporaryDirectory("sbtTesting"))
-      .bracket { tempDir =>
+case class ServiceUnderTest(details: StarterDetails) extends FLogging {
+  import Setup._
+
+  def run(tests: Option[TestFunction]*): Unit = {
+    val log = new StringBuilder()
+    val generatedFile = ZipGenerator.service.generateZipFile(details)
+    IO
+      .blocking(BFile.newTemporaryDirectory("sbtTesting"))
+      .flatMap { tempDir =>
         for {
-          zipFile <- file.map(_.toScala)
+          zipFile <- generatedFile.map(_.toScala)
+          _ <- logStep(log, "* zip file was generated")
           _ <- IO.blocking {
             zipFile.unzipTo(tempDir)
             zipFile.delete()
           }
-          _ <- IO.blocking(os.proc("sbt", ";compile ;test").call(cwd = os.Path(tempDir.toJava)).exitCode shouldBe 0)
+          _ <- logStep(log, "* zip file was unzipped")
+          service <- IO.blocking(Service(tempDir))
+        } yield (tempDir, service)
+      }
+      .bracket { pair =>
+        {
+          val (_, service) = pair
+          for {
+            port <- service.port
+            _ <- logStep(log, s"* service compiled, tested & started on port $port")
+            _ <- tests.filter(_.isDefined).map(_.get).map(test => IO.blocking(test(port))).toList.parSequence
+            _ <- logStep(log, s"* integration tests on port $port were finished")
+          } yield ()
+        }
+      } { pair =>
+        {
+          val (tempDir, service) = pair
+          for {
+            _ <- IO.blocking(service.close())
+            _ <- IO.blocking(tempDir.delete())
+          } yield ()
+        }
+      }
+      .timeoutAndForget(FiniteDuration(90L, SECONDS))
+      .onError { e =>
+        Assertions.fail(
+          s"Only the following test steps were finished for configuration '${details.describe}':$log\n${describe(e)}"
+        )
+      }
+      .unsafeRunSync()
+  }
 
-        } yield zipFile
+  private def logStep(steps: mutable.StringBuilder, step: String): IO[Unit] = IO { val _ = steps.append('\n').append(step) }
 
-      }(tempDir => IO.blocking(tempDir.delete()))
-      .unsafeRunSync()
+  private def describe(e: Throwable) = {
+    Using.Manager { _ =>
+      val sw = new StringWriter()
+      val pw = new PrintWriter(sw)
+      e.printStackTrace(pw)
+      sw.toString
+    }.get
+  }
+}
+
+case class Service(tempDir: better.files.File) extends FLogging {
+  private val portPattern = new Regex("^(?:\\[info\\] )(?:Go to |Server started at )http://localhost:(\\d+).*")
+
+  private val process: SubProcess = {
+    os.proc(
+      "sbt",
+      "-no-colors",
+      "set run / fork := true",
```

Comment:
Add comments like: 
//  forward std input to forked process https://www.scala-sbt.org/1.x/docs/Forking.html#Configuring+Input